### ISSUE TITLE
verit tactic fails with "Uncaught exception"

### ISSUE THREAD

**Issue by MrSet** [smtcoq/smtcoq#45]
I wanted to test smt-coq for coq 8.9, but I am having trouble getting it to run. 

I am on the `v8.9` (also tried out master) branch, and are using the image of `veriT` provided in `install.md`. All of the tools compiles fine (tested both with ocaml 4.04.0 and 4.07.0) but when I test the  `Example.v` does all of the examples involving `verit` fail with a message similar to: 

```
Error:
In nested Ltac calls to "verit_bool (constr_list)" and "verit_bool_base (constr_list)", last call failed.
Anomaly
"Uncaught exception Failure("Verit.call_verit: command veriT --proof-prune --proof-merge --proof-with-sharing --cnf-definitional --disable-ackermann --input=smtlib2 --proof=/var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/verit_coqdf0899.vtlog /var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/verit_coqdf0899.smt2 2> /var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/warnings_verit1e3d63.log exited with code 11")."
Please report at http://coq.inria.fr/bugs/.
```

I can't figure out what the problem is, as I can produce proofs with veriT on the command line using the command provided in the `readme.md` (except the `--disable-e` command which isn't valid?). 
```
veriT --proof-prune --proof-merge --proof-with-sharing --cnf-definitional --disable-e --disable-ackermann --input=smtlib2 --proof=file.log file.smt2
```

I have tried to add the `--verbose` flag to `call_verit` but it doesn't provide any output. Does anyone have suggestions for what the problem might be?

**Comment by ckeller** [comment]
Hi @MrSet 

Sorry for this error. It may be that directories are invalid. Can you try the whole command:
```
veriT --proof-prune --proof-merge --proof-with-sharing --cnf-definitional --disable-ackermann --input=smtlib2 --proof=/var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/verit_coqdf0899.vtlog file.smt2 2> /var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/warnings_verit1e3d63.log
```

**Comment by MrSet** [comment]
Thanks for the suggestion. 

It doesn't seem like the directory is wrong, as when I execute the command which fails inside coq on the command line, it seems to work. For example:  
```
veriT --proof-prune --proof-merge --proof-with-sharing --cnf-definitional --disable-ackermann --input=smtlib2 --proof=/var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/verit_coqdf0899.vtlog /var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/verit_coqdf0899.smt2 2> /var/folders/_r/mjtfy67x5jx9jw8h7959j2c80000gn/T/warnings_verit1e3d63.log
```

I can both find the expected output files in the directory and it prints the following output in the terminal: 

```
veriT dev - the SMT-solver veriT (UFRN/LORIA).
success
success
success
success
success
unsat
success
success
```

**Comment by ckeller** [comment]
When you observe an error from Coq, is the file "/var/folders/_r/***/T/warnings_verit***.log" created? If so, can you copy-paste it?

**Comment by MrSet** [comment]
The `warnings_verit.log` file is unfortunately not created. However, is both the `.vtlog` and the `.smt2` created. Containing respectively (for the first call to `verit_bool` in `Example.v`): 
```
1:(input (#1:(and #2:(or op_0 op_2 op_1) #3:(or (not op_0) (not op_2) (not op_1)) #4:(or (not op_0) op_2) #5:(or (not op_2) op_1) #6:(or (not op_1) op_0))))
2:(and (#2) 1 0)
3:(or (op_0 op_2 op_1) 2)
4:(and (#3) 1 1)
5:(or ((not op_0) (not op_2) (not op_1)) 4)
6:(and (#4) 1 2)
7:(or ((not op_0) op_2) 6)
8:(and (#5) 1 3)
9:(or ((not op_2) op_1) 8)
10:(and (#6) 1 4)
11:(or ((not op_1) op_0) 10)
12:(resolution (op_0) 3 9 11)
13:(resolution (op_2) 7 12)
14:(resolution ((not op_1)) 5 13 12)
15:(resolution () 9 13 14)
```
and 
```
(set-logic UFLIA)
(declare-fun op_1 () Bool)
(declare-fun op_0 () Bool)
(declare-fun op_2 () Bool)
(assert (and (or op_0 op_2 op_1) (or (not op_0) (not op_2) (not op_1)) (or (not op_0) op_2) (or (not op_2) op_1) (or (not op_1) op_0)))
(check-sat)
(exit)
```

**Comment by ckeller** [comment]
Well, it seems that verit works correctly but returns a non-zero exit code (?), and SMTCoq fails due to it.

Can you pull master and try again? Thanks

**Comment by MrSet** [comment]
Thanks for the try. But unfortunately, this didn't fix it. Instead, I got: 
```
Error:
In nested Ltac calls to "verit_bool (constr_list)" and "verit_bool_base (constr_list)", last call failed.
veriT returns 'unknown'
```

In the same file as you commented out the check for exit codes I did, however, find the reason for why no warnings were produced: The warning file is explicitly deleted when an assertion fails. I tried to comment out this specific deletion of the warning file (L196 in `verit.ml`). The content of the `warning_veritxxxxxx.log` is: 

```
Invalid memory reference
```

**Comment by MrSet** [comment]
Ahh! Finally some progress! If I also comment out the check for that warnings are empty, then I can actually run the `verit` and `verit_bool` tactic!

**Comment by ckeller** [comment]
Thanks, that can be a quick fix, but I would like to understand why verit warns about an invalid memory reference...

**Comment by MrSet** [comment]
So would I! I will try to investigate things further later. Also: Seems like this didn't solve all my problems. When using `smt` for the bitvector examples in `Example.v` I get a: 
```
Ltac call to "smt (constr_list)" failed.
Anomaly "Uncaught exception Failure("VeritSyntax.get_clause : clause number -1 not found\n")."
```

**Comment by ckeller** [comment]
Is your architecture x86_64?

**Comment by MrSet** [comment]
Yes it is

**Comment by MrSet** [comment]
Apparently, is it only one of the bitvector examples, namely: 
```
Goal forall (bv1 bv2 bv3: bitvector 4),
    bv1 = #b|0|0|0|0|  /\
    bv2 = #b|1|0|0|0|  /\
    bv3 = #b|1|1|0|0|  ->
    bv_ultP bv1 bv2 /\ bv_ultP bv2 bv3.
Proof.
  smt.
Qed.
```
That fails with the following exception: 
```
Ltac call to "smt (constr_list)" failed.
Anomaly "Uncaught exception Failure("VeritSyntax.get_clause : clause number -1 not found\n")."
```
I suspect that this is just the error message that occurs when the tactic fails? If so, is this not supposed to be an example that goes through?

**Comment by ckeller** [comment]
No, this example should go through (and it does on my machine). Clauses are indexed starting from 0. It may be related to the memory problem of veriT but I have no idea...

**Comment by ckeller** [comment]
@vblot suggests that you send us the file vtlog generated by veriT, can you do so?

Anyway, the two problems make me think of something related to your architecture/system... Can you tell us more about it?

**Comment by MrSet** [comment]
Sure, thanks for taking the time to investigate it!

I just found out that if I run the example: 
```
Goal forall (bv1 bv2 bv3: bitvector 4),
    bv1 = #b|0|0|0|0|  /\
    bv2 = #b|1|0|0|0|  /\
    bv3 = #b|1|1|0|0|  ->
    bv_ultP bv1 bv2 /\ bv_ultP bv2 bv3.
Proof.
  smt.
Qed.
```
With the cvc4 tactic rather than the smt/verit tactics, it will succeed. Can you investigate what happens if you run that example with either only verit/cvc4? If it indeed only is cvc4 which solves the goal for you as well, then the error might still just be something about the error-handling from verit. 

I am running it on, (what I suppose) is a fairly standard system: 
```
MacBook Pro (15-inch, 2018)
OS: Mojave 10.14.3
Processor: Intel Core i7 2,2 GHz, 6 cores (I7-8750H)
Memory: 16 GB 2400 MHz DDR4
```
Anything else you will like to know?

The `.vtlog` file that is produced when trying the above example is **empty**.

The `warning_veritXXXXXX.log` file has the following content: 
```
warning : proof_done: status is still open
Invalid memory reference
```

The `.smt2` file has the following content: 
```
(set-logic UFLIA)
(declare-sort Tindex_0 0)
(declare-fun op_2 () Tindex_0)
(declare-fun op_1 () Tindex_0)
(declare-fun op_3 () Tindex_0)
(declare-fun op_5 () Tindex_0)
(declare-fun op_4 () Tindex_0)
(declare-fun op_6 (Tindex_0 Tindex_0) Bool)
(declare-fun op_0 () Tindex_0)
(assert (not (=> (and (= op_4 op_5) (= op_2 op_3) (= op_0 op_1)) (and (op_6 op_4 op_2) (op_6 op_2 op_0)))))
(check-sat)
(exit)
```

**Comment by ckeller** [comment]
You are right: the "smt" tactic tries verit first, but since you remove the check of warnings (which is the way veriT tells that it did not find a proof), you do not fail early enough.

So I need to understand why verit returns this error on your OS. I will try to get a Mac to experiment with.

Unfortunately, I am using an old version of veriT because the certificate format as changed and I did not find time to switch to a recent version yet... I will see how long it would take to do so.

**Comment by MrSet** [comment]
I see. By adding the following very ugly check  to `assert_empty_warnings` in `verit.ml` I can now finally use the `smt` tactic for all cases in `Example.v`: 
```
  let assert_empty_warnings () =
    try
      while true do
        let l = input_line win in 
        if l != "Invalid memory reference"
        then raise Non_empty_warnings
      done
    with End_of_file -> () in
```

Now what is left is only to figure out why the "Invalid memory reference" occurs in the first place...

**Comment by ckeller** [comment]
Thanks to you for the report and your tests!

I changed a bit your fix to output warnings, can you try the branch "issue45"? https://github.com/smtcoq/smtcoq/tree/issue45

Before interacting with veriT developers, I need to switch to a more recent version of veriT. It is on my TODO list https://github.com/smtcoq/smtcoq/issues/18 . When it is done, do you agree to test again?

**Comment by MrSet** [comment]
Branch `issue45` works perfectly fine for me. It outputs both a warning for "exit code 11" and one for "Invalid memory reference" which I guess is the intended behavior.  

Sure, you can just write to me or comment on this issue whenever you want me to test something.

**Comment by ckeller** [comment]
Great, thanks!

I hope you will enjoy SMTCoq
